### PR TITLE
Add `UnnecessaryRun` rule

### PR DESCRIPTION
**Pull request by atulgpt**
This rule takes inspiration from `UnnecessaryApply` rule

Fixes https://github.com/detekt/detekt/issues/6101

### CODE REVIEW

**Comment by atulgpt** [comment]
@cortinico @BraisGabin should we merge this PR?

File: detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryRun.kt [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+        (
+            lambdaArguments.firstOrNull()?.getLambdaExpression()
+                ?: valueArguments
+                    .firstOrNull()
+                    ?.getArgumentExpression()
+                    ?.unpackFunctionLiteral()
+            )
+}
+
+private val RUN_FQ_NAME = FqName("kotlin.run")
```
Comment (cortinico) [inline detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryRun.kt:124]: Can't this live inside the `UnnecessaryRun` class?

**Review by cortinico — APPROVED** [review]
LGTM 👍

File: detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryRun.kt (hunk 1 above)
Comment (atulgpt) [reply detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryRun.kt:124]: That will create new instance of `RUN_FQ_NAME` for every rull instance. Or will need the companion object, right?

Comment (atulgpt) [reply detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryRun.kt:124]: `UnnecessaryApply` also has same style where consts are defined as private file level variables

Comment (cortinico) [reply detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryRun.kt:124]: Yup let's keep it there then for consistency with `UnnecessaryApply`

**Comment by atulgpt** [comment]
Hi, @BraisGabin / @cortinico should we merge this PR? It's approved and open for very long time

**Comment by cortinico** [comment]
Can we get a review from another @detekt/maintainers ?

**Review by schalkms — COMMENTED** [review]
I'm afraid, I'm pretty busy these days, so can't review it thoroughly.
But, as this is pretty similar to the `UnnecessaryApply` rule, it should be good to go.

**Comment by atulgpt** [comment]
Hi, @cortinico should we merge this PR since it follows the existing rule. If there is any issue we can improve around the same.(Keeping this open requires constant rebase from the main 😅)

File: detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryRunSpec.kt [hunk 2]
```diff
@@ -0,0 +1,493 @@
...
+                                prop = 1
+                            }
+
+                            val a3 = c.run {
+                                prop += 1
+                            }
+
+                            val a4 = c.run {
+                                prop /= 1
+                            }
```
Comment (TWiStErRob) [inline detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryRunSpec.kt:381]: All these `a*` are `: Unit` right? because the assignment expression is not like Java.

File: detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryRunSpec.kt [hunk 3]
```diff
@@ -0,0 +1,493 @@
...
+                        }
+                    """.trimIndent()
+                )
+            ).hasSize(3)
+        }
+
+        @Test
+        fun `does not report when result is used`() {
```
Comment (TWiStErRob) [inline detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryRunSpec.kt:357]: feels like a similar test is needed for using the value in different ways, like

```kotlin
f(c.run { prop = 1 })
```

```kotlin
a + c.run { prop = 1 }
```

```kotlin
"${c.run { prop = 1 })}"
```

File: detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryRunSpec.kt [hunk 4]
```diff
@@ -0,0 +1,493 @@
...
+import io.gitlab.arturbosch.detekt.api.Config
+import io.gitlab.arturbosch.detekt.rules.KotlinCoreEnvironmentTest
+import io.gitlab.arturbosch.detekt.test.assertThat
+import io.gitlab.arturbosch.detekt.test.compileAndLintWithContext
+import org.jetbrains.kotlin.cli.jvm.compiler.KotlinCoreEnvironment
+import org.junit.jupiter.api.Nested
+import org.junit.jupiter.api.Test
+
+@KotlinCoreEnvironmentTest
+class UnnecessaryRunSpec(val env: KotlinCoreEnvironment) {
```
Comment (TWiStErRob) [inline detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryRunSpec.kt:12]: did you consider this edge case, which is near impossible to fix? (apart from changing to other scope functions)

```kotlin
interface Foo {
   fun Bar.baz()
}
fun f(foo: Foo, bar: Bar) {
    foo.run { bar.baz() }
}
```

File: detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryRunSpec.kt (hunk 2 above)
Comment (atulgpt) [reply detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryRunSpec.kt:381]: `All these a* are : Unit right?` yes all these are `Unit`

File: detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryRunSpec.kt (hunk 4 above)
Comment (atulgpt) [reply detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryRunSpec.kt:12]: Yes, this is failing. Problem here is to find the places where scope is used(traverse the AST nodes filter-out the applicable usages and then check if scope if used and if that is same as provided by the `run`)

Adding the complete TC for ref
```kotlin
@Test
fun `do not report when scope by run is used`() {
    assertThat(
        subject.compileAndLintWithContext(
            env,
            """
                class Bar
                interface Foo {
                    fun Bar.baz()
                }
                fun f(foo: Foo, bar: Bar) {
                    foo.run { bar.baz() }
                }
            """.trimIndent()
        )
    ).isEmpty()
}
```

File: detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryRunSpec.kt (hunk 3 above)
Comment (atulgpt) [reply detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryRunSpec.kt:357]: Add these TC in this commit https://github.com/detekt/detekt/pull/7428/commits/2e6245ba658b6de2432de005b559c5c5ab8de731

File: detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryRunSpec.kt (hunk 2 above)
Comment (TWiStErRob) [reply detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryRunSpec.kt:381]: Makes sense, it's just a bit weird code. Maybe it's not this rule's job to highlight it though. There could be more realistic examples, like: `fun f(): Unit = c.run { prop += 1 }`

File: detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryRunSpec.kt (hunk 4 above)
Comment (TWiStErRob) [reply detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryRunSpec.kt:12]: ~I guess we can create a bug and merge this?~ I wonder how often this would be the case, as `apply` and `with` are usually more applicable to this use case (not returning the value), so it might not even be a false positive?

File: detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryRunSpec.kt (hunk 2 above)
Comment (atulgpt) [reply detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryRunSpec.kt:381]: I can also add `fun f(): Unit = c.run { prop += 1 }` to this example. What do you say?

File: detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryRunSpec.kt (hunk 4 above)
Comment (atulgpt) [reply detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryRunSpec.kt:12]: Agree but TBH when rule will be used out by many other devs then we will find more cases where these type patterns might become useful or require changes in the rule

File: detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryRunSpec.kt (hunk 2 above)
Comment (atulgpt) [reply detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryRunSpec.kt:381]: Hi @TWiStErRob added the TC at https://github.com/detekt/detekt/pull/7428/commits/8b5e94421702ac1b2713fd0c48d975f1176d8dda

File: detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryRunSpec.kt (hunk 4 above)
Comment (atulgpt) [reply detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryRunSpec.kt:12]: @TWiStErRob do you agree?

File: detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryRun.kt [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+ * config.version = "1.2"
+ * config?.environment = "test"
+ * val final = getNullableValue() ?: fallback
+ * stringBuilder.run {
+ *     append("test")
+ *     toString()
+ * }
+ * </compliant>
+ */
+@RequiresFullAnalysis
```
Comment (3flex) [inline detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryRun.kt:47]: Apologies but we are migrating to the Analysis API so it would be much better if this can be implemented using that instead, otherwise we'll have to migrate it

Comment (schalkms) [reply detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryRun.kt:47]: @atulgpt do you still plan to push this PR forward by leveraging the new Analysis API? Or shall we close this PR?

Comment (atulgpt) [reply detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryRun.kt:47]: Hi @schalkms I am closing this PR as of now.. Will reopen it as some of TCs which were possible earlier are now not possible with AA apis

Comment (schalkms) [reply detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryRun.kt:47]: Ok, good to know. Ty for the reply.
